L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc

```
@@ -0,0 +1,134 @@
+
+    for(; l1tTriggerTowerMappingStream >> trigger_cell_id >> iEta >> iPhi;) {
+      if(iEta >= nBinsEta_ || iPhi >= nBinsPhi_) {
+        throw edm::Exception(edm::errors::Configuration, "Configuration")
+          << "HGCalTriggerTowerGeometryHelper warning inconsistent mapping TC : " << trigger_cell_id
+          << " to TT iEta: " << iEta << " iPhi: " << iPhi
+          << " when max #bins eta: "  << nBinsEta_ << " phi: " << nBinsPhi_ << std::endl;
+      }
+      HGCalDetId detId(trigger_cell_id);
+      int zside = detId.zside();
+      l1t::HGCalTowerID towerId(zside, iEta, iPhi);
+      cells_to_trigger_towers_[trigger_cell_id] = towerId.rawId();
+    }
+    l1tTriggerTowerMappingStream.close();
+
+  }
+}
+
+
+const std::vector<l1t::HGCalTowerCoord>& HGCalTriggerTowerGeometryHelper::getTowerCoordinates() const {
+  return tower_coords_;
+}
+
+
+unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTowerFromTriggerCell(const unsigned trigger_cell_id, const float& eta, const float& phi) const {
+  // NOTE: if the TC is not found in the map than it is mapped via eta-phi coords.
+  // this can be considered dangerous (silent failure of the map) but it actually allows to save
+  // memory mapping explicitly only what is actually needed
+  auto tower_id_itr = cells_to_trigger_towers_.find(trigger_cell_id);
+  if(tower_id_itr != cells_to_trigger_towers_.end()) return tower_id_itr->second;
+
+  auto bin_eta_l = std::lower_bound(binsEta_.begin(), binsEta_.end(), fabs(eta));
+  unsigned int bin_eta = 0;
+  // we add a protection for TCs in Hadron part which are outside the boundaries and possible rounding effects
+  if(bin_eta_l == binsEta_.end()) {
+    if(fabs(eta) < minEta_) {
+      bin_eta = 0;
+    } else if(fabs(eta) >= maxEta_) {
+      bin_eta = nBinsEta_;
+    } else {
+      edm::LogError("HGCalTriggerTowerGeometryHelper") << " did not manage to map TC " << trigger_cell_id << " (eta: " << eta << ") to any Trigger Tower\n";
```

should it throw instead of just printing an error?

---

@kpedro88 I think that the error is ok for now. Maybe we will review this decision once we have some more experience with this code.